src/ray/object_manager/pull_manager.cc

```diff
@@ -39,33 +42,183 @@ uint64_t PullManager::Pull(const std::vector<rpc::ObjectReference> &object_ref_b
+                     << next_request_it->first;
+      return false;
+    }
+  }
+
+  // Activate the bundle.
+  for (const auto &ref : next_request_it->second) {
+    auto obj_id = ObjectRefToId(ref);
+    bool start_pull = active_object_pull_requests_.count(obj_id) == 0;
+    active_object_pull_requests_[obj_id].insert(next_request_it->first);
+    if (start_pull) {
+      RAY_LOG(DEBUG) << "Activating pull for object " << obj_id;
+      // This is the first bundle request in the queue to require this object.
+      // Add the size to the number of bytes being pulled.
+      // NOTE(swang): The size could be 0 if we haven't received size
+      // information yet. If we receive the size later on, we will update the
+      // total bytes being pulled then.
+      auto it = object_pull_requests_.find(obj_id);
+      RAY_CHECK(it != object_pull_requests_.end());
+      num_bytes_being_pulled_ += it->second.object_size;
+
+      object_ids_to_pull->insert(obj_id);
+    }
+  }
+
+  // Update the pointer to the last pull request that we are actively pulling.
+  RAY_CHECK(next_request_it->first > highest_req_id_being_pulled_);
+  highest_req_id_being_pulled_ = next_request_it->first;
+  return true;
+}
+
+void PullManager::DeactivatePullBundleRequest(
+    std::map<uint64_t, std::vector<rpc::ObjectReference>>::iterator request_it,
+    std::unordered_set<ObjectID> *object_ids_to_cancel) {
+  for (const auto &ref : request_it->second) {
+    auto obj_id = ObjectRefToId(ref);
+    RAY_CHECK(active_object_pull_requests_[obj_id].erase(request_it->first));
+    if (active_object_pull_requests_[obj_id].empty()) {
+      RAY_LOG(DEBUG) << "Deactivating pull for object " << obj_id;
+      auto it = object_pull_requests_.find(obj_id);
+      RAY_CHECK(it != object_pull_requests_.end());
+      num_bytes_being_pulled_ -= it->second.object_size;
+      active_object_pull_requests_.erase(obj_id);
+      if (object_ids_to_cancel) {
+        object_ids_to_cancel->insert(obj_id);
+      }
+    }
+  }
+
+  // If this was the last active request, update the pointer to its
+  // predecessor, if one exists.
+  if (highest_req_id_being_pulled_ == request_it->first) {
+    if (request_it == pull_request_bundles_.begin()) {
+      highest_req_id_being_pulled_ = 0;
+    } else {
+      highest_req_id_being_pulled_ = std::prev(request_it)->first;
+    }
+  }
+}
+
+void PullManager::UpdatePullsBasedOnAvailableMemory(size_t num_bytes_available) {
+  // TODO(swang): We could return an error for requests whose total size is
+  // larger than the capacity of the memory store. This would hang right now.
+
+  if (num_bytes_available_ != num_bytes_available) {
+    RAY_LOG(DEBUG) << "Updating pulls based on available memory: " << num_bytes_available;
+  }
+  num_bytes_available_ = num_bytes_available;
+
+  std::unordered_set<ObjectID> object_ids_to_pull;
+  // While there is available capacity, activate the next pull request.
```

Comment:
Should we always pull at least the first bundle in the queue? Otherwise I can see the workload stalling when it is possible we can make space via object spilling.

Reply:
Hmm that's a good point, I'll change that and see if I can write a regression test for it.

Reply:
Actually, it turns out this doesn't work because the object store just keeps evicting other objects needed by the first bundle, so it never triggers the OOM handling. So I'm going to change this to directly trigger OOM instead.